src/language/records.md

```diff
@@ -0,0 +1,160 @@
+---
+title: Records
+description: Summary of the record data structure in Dart.
+toc: false
+---
+
+Records are an anonymous, immutable, aggregate type. Like other [collection types][], 
+they let you bundle multiple objects into a single object. Unlike other collection 
+types, records are fixed-sized, heterogeneous, and typed.
+
+Records are real values; you can store them in variables, pass them to and from
+functions, and store them in lists.
+
+## Record syntax
+
+_Records expressions_ are comma-delimited lists of named or positional fields,
+enclosed in parenthesis:
+
+```dart
+var record = ('first', a: 2, b: true, 'last');    // Record expression
+```
+
+_Record type annotations_ are comma-delimited lists of types enclosed in parenthesis.
+The `(int, int)` statements here are record type annotations used as a return
+type and as parameter types:
+
+```dart
+(int, int) swap((int, int) record) {
+  var (a, b) = record;
+  return (b, a);
+}
+```
+
+Named fields inside _record type annotations_ go inside a brace-delimited section
+of type and name pairs, after all positional fields:
+
+```dart
+// Record type annotation in a variable declaration
+(String, String?, {int a, bool b}) record;
+```
+
+Naming fields in record type annotations is optional for documentation purposes.
+The names of those named fields are not optional, though,  
+once you create a record with named fields. The field names become part of the
+[record's type definition](#record-types). They can be omitted, but not replaced.
```

Comment:
**issue:** Sentence structure per [GSG](https://developers.google.com/style/sentence-structure)
```suggestion
Once you create a record with named fields,
those field names become part of the [record's type definition](#record-types).
They can be omitted, but not replaced.
```

**question:**  If they're not optional, how can you omit the field names?

Reply:
It's confusing. How I understand it, which I haven't found a good way to put into words yet, is: once you name fields and actually create a record using those names (so, not just using a record type annotation to define its shape, but using a record expression to populate the fields), the names become part of the record's definition, i.e. it's shape. 

You can create a record from an annotation with named fields _without_ using the names, because they're implicit(maybe? not super sure about this part), but you _can't_ create a record from an annotation with named fields using different names in the expression...

```dart 
   ({int x, int y}) someRecord = (x: 1, y: 2);
// ^^^^^^^^^^^^^^^              ^^^^^^^^^^^^
// type annotation              expression

   ({int x, int y}) someRecord;
   someRecord = (1, 2);
// fine, field values fit the types in the annotation, and don't clash with the names
   print(someRecord.x);    // prints 1, i think?
   print(someRecord.y);    // prints 2, i think?

   ({int x, int y}) someRecord;
   someRecord = (a: 1, b: 2);
// doesn't work, can't give the fields different names because someRecord's shape is defined as named int x and named int y
```

So, something like that... but I'm not 100% sure that's accurate

Reply:
I'll see if I can help:

You can think of every record as having two separate sets of fields:

* A set of positional fields.
* A set of named fields.

These two sets are totally independent of each other: a named field isn't a positional field and a positional field isn't a named field. Either set may be empty, or they can both be. You can have records with just positional fields, just named, both, or no fields at all.

### Record expressions

When you write a record expression to create a record object, the positional fields are written as simple expressions:

```dart
var record = (1, 2, 3);
//          ^  ^  ^ Positional fields.
```

The named fields have a name and `:` before the field value:

```dart
var record = (x: 1, y: 2, z: 3);
//          ^^^^  ^^^^  ^^^^ Named fields.
```

You can have both in a single record expression and they can be arbitrarily
interleaved:

```dart
var record = (x: 1, 2, y: 3, 4);
//                ^        ^  Positional fields.
//          ^^^^     ^^^^  Named fields.
```

### Record types

Every record object has a static type, which is how the type checker "sees" the record. Users write down these types using "type annotations" (which are not related to metadata annotations, the `@foo` syntax).

In a record type annotation, the positional fields are written as simple type annotations for the field types:

```dart
  (int, String) record = (1, 'str');
// ^^^  ^^^^^^ Positional fields.
```

Named fields are enclosed in curly braces and must have a name after the type:

```dart
  ({int a, String b}) record = (a: 1, b: 'str');
//  ^^^^^  ^^^^^^^^ Named fields.
```

If a record type annotation has both positional fields and named ones, the positional ones must come first:

```dart
  (double, bool, {int a, String b}) record = (1.2, true, a: 1, b: 'str');
// ^^^^^^^ ^^^^ Positional fields.
//                ^^^^^  ^^^^^^^^ Named fields.
```

### Positional field names

The above is, I hope, fairly clear and straightforward. Where it gets a little confusing is that we do allow names on positional fields *in record type annotations*:

```dart
  (int x, String y) record = (1, 'str');
// ^^^^^  ^^^^^^^^ Positional fields.
```

Note that the fields aren't enclosed in curly braces. These are still positional fields and are not named fields. Note how in the record expression in the initializer, there are no names. It's just that the fields in the type annotation have, uh, "names".

But these "names" are purely for documentation and have no meaning to the language or the type checker. It's essentially the same as doing:

```dart
  (int /* x */, String /* y */) record = (1, 'str');
// ^^^          ^^^^^^ Positional fields.
```

And, in particular, these names are *not* part of the record's actual type:

```dart
(int x, String y) record1 = (1, 'str');
(int a, String b) record2 = record1; // OK!
```

These two variables, `record1` and `record2` have the exact same type even though the positional field "names" are different, which is why the assignment is valid.

This is different from actual named fields where the name *is* part of the type:

```dart
({int x, String y}) record1 = (x: 1, y: 'str');
({int a, String b}) record2 = record1; // Error!
```

This assignment is invalid because `record1` and `record2` have different types. Note also that the record expression in the initializer now says the field names.

The proposal sometimes talks about the "name" of a positional field, but really we should have a different term for it because it's not the same as an actual named field's name.

Sorry this is confusing. Much of this comes from deliberately designing records to mirror function parameter lists, so we inherit some weirdness from that. But I think the overall correspondence is useful (especially if we ever add the ability to spread a record into an argument list).

You can see the exact same confusion around positional parameter names in function typedefs:

```dart
typedef Positional1 = Function(int x, String y);
typedef Positional2 = Function(int a, String b);
```

These two typedefs define equivalent types even though the positional parameters have different names. But these two are different types from each other:

```dart
typedef Named1 = Function({int x, String y});
typedef Named2 = Function({int a, String b});
```